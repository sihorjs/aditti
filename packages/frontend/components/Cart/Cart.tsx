import React from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";

import {
    increaseItemQuantity,
    decreaseItemQuantity,
    clearCart,
    removeItem
} from "actions/cart";
import { getCartTotalPrice, getCartItems } from "selectors/cart";

import State from "types/state";
import { CartProductType } from "@aditti/types";

type Props = {
    cartItems: Array<CartProductType>;
    onClearCart: () => void;
    onDecreaseQuantity: (id: string) => void;
    onIncreaseQuantity: (id: string) => void;
    onRemoveItem: (id: string) => void;
    totalPrice: number;
};

export function Cart(props: Props) {
    return props.totalPrice ? (
        <div className="cart">
            <div className="cart__title">Your bucket</div>
            <div className="cart__items">
                {props.cartItems.map(item => {
                    return (
                        <div key={item.id}>
                            <div>
                                <button
                                    className="
                                            cart__item-remove
                                            flaticon-cancel
                                        "
                                    onClick={() => props.onRemoveItem(item.id)}
                                    type="button"
                                />
                                <button
                                    className="
                                            cart__item-decrease
                                            flaticon-minus
                                        "
                                    onClick={() =>
                                        props.onDecreaseQuantity(item.id)
                                    }
                                    type="button"
                                />
                                <img alt={item.name} src={item.src} />
                                <button
                                    className="
                                            cart__item-increase
                                            flaticon-plus
                                        "
                                    onClick={() =>
                                        props.onIncreaseQuantity(item.id)
                                    }
                                    type="button"
                                />
                            </div>
                            <div>{item.name}</div>
                            <div>
                                <span>{item.quantity}</span>
                                <span>* {item.price}</span>
                                <span>
                                    {` = ${item.quantity * item.price}`}
                                </span>
                            </div>
                        </div>
                    );
                })}
            </div>
            <div className="cart__footer">
                <span>{`Your total sum is ${props.totalPrice}`}</span>
                <button
                    className="cart__clear"
                    onClick={props.onClearCart}
                    type="button"
                >
                    Clear bucket
                </button>
            </div>
        </div>
    ) : (
        <div className="empty-cart">
            <div className="empty-cart__title">Your bucket is empty</div>
        </div>
    );
}

function mapStateToProps(state: State) {
    return {
        totalPrice: getCartTotalPrice(state),
        cartItems: getCartItems(state)
    };
}

function mapDispatchToProps(dispatch: Dispatch) {
    return {
        onIncreaseQuantity: (id: string) => dispatch(increaseItemQuantity(id)),
        onDecreaseQuantity: (id: string) => dispatch(decreaseItemQuantity(id)),
        onRemoveItem: (id: string) => dispatch(removeItem(id)),
        onClearCart: () => dispatch(clearCart())
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Cart);
