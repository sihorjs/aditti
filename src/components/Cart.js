import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import {
    increaseItemQuantity,
    decreaseItemQuantity,
    clearCart,
    removeItem
} from "actions/cart";
import { getCartTotalPrice, getCartItems } from "selectors/cart";

class Cart extends Component {
    render() {
        return this.props.totalPrice ? (
            <div className="dialog">
                <h2 className="dialog-header">Your bucket</h2>
                <div className="dialog-main">
                    {this.props.cartItems.map((item, index) => {
                        if (item.quantity) {
                            return (
                                <div className="item-row" key={index}>
                                    <div className="item-controls">
                                        <button
                                            className="flaticon-cancel"
                                            onClick={() =>
                                                this.props.onRemoveItem(item.id)
                                            }
                                        />
                                        <button
                                            className="flaticon-minus"
                                            onClick={() =>
                                                this.props.onDecreaseQuantity(
                                                    item.id
                                                )
                                            }
                                        />
                                        <img src={item.src} alt={item.name} />
                                        <button
                                            className="flaticon-plus"
                                            onClick={() =>
                                                this.props.onIncreaseQuantity(
                                                    item.id
                                                )
                                            }
                                        />
                                    </div>
                                    <div className="item-name">{item.name}</div>
                                    <div className="item-total">
                                        <span>{item.quantity}</span>
                                        <span>* {item.price}</span>
                                        <span>
                                            {` = ${item.quantity * item.price}`}
                                        </span>
                                    </div>
                                </div>
                            );
                        }
                    })}
                </div>
                <div className="dialog-footer">
                    <span>`Your total sum is ${this.props.totalPrice}`</span>
                    <button onClick={() => this.props.onClearBucket()}>
                        Clear bucket
                    </button>
                </div>
            </div>
        ) : (
            <div className="dialog empty-dialog">
                <h2 className="dialog-header">Your bucket is empty</h2>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {
        totalPrice: getCartTotalPrice(state),
        cartItems: getCartItems(state)
    };
}

function mapDispatchToProps(dispatch) {
    return {
        onIncreaseQuantity: id => dispatch(increaseItemQuantity(id)),
        onDecreaseQuantity: id => dispatch(decreaseItemQuantity(id)),
        onRemoveItem: id => dispatch(removeItem(id)),
        onClearCart: () => dispatch(clearCart())
    };
}

Cart.propTypes = {
    cartItems: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            price: PropTypes.number.isRequired,
            category: PropTypes.string.isRequired,
            src: PropTypes.string.isRequired
        })
    ).isRequired,
    onClearBucket: PropTypes.func.isRequired,
    onDecreaseQuantity: PropTypes.func.isRequired,
    onIncreaseQuantity: PropTypes.func.isRequired,
    onRemoveItem: PropTypes.func.isRequired,
    totalPrice: PropTypes.number.isRequired
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Cart);
