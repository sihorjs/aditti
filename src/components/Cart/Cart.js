import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import {
    increaseItemQuantity,
    decreaseItemQuantity,
    clearCart,
    removeItem
} from "actions/cart";
import { getCartTotalPrice, getCartItems } from "selectors/cart";

export function Cart(props) {
    return props.totalPrice ? (
        <div className="cart">
            <div className="cart__title">Your bucket</div>
            <div className="cart__items">
                {props.cartItems.map(item => {
                    return (
                        <div key={item.id}>
                            <div>
                                <button
                                    className="
                                            cart__item-remove
                                            flaticon-cancel
                                        "
                                    onClick={() => props.onRemoveItem(item.id)}
                                    type="button"
                                />
                                <button
                                    className="
                                            cart__item-decrease
                                            flaticon-minus
                                        "
                                    onClick={() =>
                                        props.onDecreaseQuantity(item.id)
                                    }
                                    type="button"
                                />
                                <img src={item.src} alt={item.name} />
                                <button
                                    className="
                                            cart__item-increase
                                            flaticon-plus
                                        "
                                    onClick={() =>
                                        props.onIncreaseQuantity(item.id)
                                    }
                                    type="button"
                                />
                            </div>
                            <div>{item.name}</div>
                            <div>
                                <span>{item.quantity}</span>
                                <span>* {item.price}</span>
                                <span>
                                    {` = ${item.quantity * item.price}`}
                                </span>
                            </div>
                        </div>
                    );
                })}
            </div>
            <div className="cart__footer">
                <span>`Your total sum is ${props.totalPrice}`</span>
                <button
                    className="cart__clear"
                    onClick={props.onClearCart}
                    type="button"
                >
                    Clear bucket
                </button>
            </div>
        </div>
    ) : (
        <div className="empty-cart">
            <div className="empty-cart__title">Your bucket is empty</div>
        </div>
    );
}

function mapStateToProps(state) {
    return {
        totalPrice: getCartTotalPrice(state),
        cartItems: getCartItems(state)
    };
}

function mapDispatchToProps(dispatch) {
    return {
        onIncreaseQuantity: id => dispatch(increaseItemQuantity(id)),
        onDecreaseQuantity: id => dispatch(decreaseItemQuantity(id)),
        onRemoveItem: id => dispatch(removeItem(id)),
        onClearCart: () => dispatch(clearCart())
    };
}

Cart.propTypes = {
    cartItems: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            price: PropTypes.number.isRequired,
            category: PropTypes.string.isRequired,
            src: PropTypes.string.isRequired
        })
    ).isRequired,
    onClearCart: PropTypes.func.isRequired,
    onDecreaseQuantity: PropTypes.func.isRequired,
    onIncreaseQuantity: PropTypes.func.isRequired,
    onRemoveItem: PropTypes.func.isRequired,
    totalPrice: PropTypes.number.isRequired
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Cart);
