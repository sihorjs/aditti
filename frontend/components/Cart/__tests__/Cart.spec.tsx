import React from "react";
import { shallow } from "enzyme";
import { Cart } from "../Cart";

describe("<Cart />", () => {
    const onIncreaseQuantity = jest.fn();
    const onDecreaseQuantity = jest.fn();
    const onRemoveItem = jest.fn();
    const onClearCart = jest.fn();

    const props = {
        onIncreaseQuantity,
        onDecreaseQuantity,
        onRemoveItem,
        onClearCart
    };

    const productPrice = 500;
    const productId = "1";
    const productQuantity = 3;

    const cartItems = [
        {
            src: "assets/img/cart",
            price: productPrice,
            name: "t-shirt",
            category: "t-shirts",
            id: productId,
            quantity: productQuantity
        }
    ];

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crash", () => {
        const tree = shallow(<Cart {...props} cartItems={[]} totalPrice={0} />);

        expect(tree).toMatchSnapshot();
    });

    it("renders empty cart if there isn't items in it", () => {
        const component = shallow(
            <Cart {...props} cartItems={[]} totalPrice={0} />
        );

        expect(component.find(".cart").exists()).toBeFalsy();
        expect(component.find(".empty-cart")).toHaveLength(1);
    });

    it("handles clear of cart", () => {
        const component = shallow(
            <Cart {...props} cartItems={cartItems} totalPrice={productPrice} />
        );

        component.find(".cart__clear").simulate("click");

        expect(onClearCart).toHaveBeenCalledTimes(1);
    });

    it("handles add to cart", () => {
        const component = shallow(
            <Cart {...props} cartItems={cartItems} totalPrice={productPrice} />
        );

        component.find(".cart__item-increase").simulate("click");

        expect(onIncreaseQuantity).toHaveBeenCalledWith(productId);
        expect(onIncreaseQuantity).toHaveBeenCalledTimes(1);
    });

    it("handles decrease item quantity in cart", () => {
        const component = shallow(
            <Cart {...props} cartItems={cartItems} totalPrice={productPrice} />
        );

        component.find(".cart__item-decrease").simulate("click");

        expect(onDecreaseQuantity).toHaveBeenCalledWith(productId);
        expect(onDecreaseQuantity).toHaveBeenCalledTimes(1);
    });

    it("handles remove item from cart", () => {
        const component = shallow(
            <Cart {...props} cartItems={cartItems} totalPrice={productPrice} />
        );

        component.find(".cart__item-remove").simulate("click");

        expect(onRemoveItem).toHaveBeenCalledWith(productId);
        expect(onRemoveItem).toHaveBeenCalledTimes(1);
    });

    it("non-empty cart has header, footer and table with products", () => {
        const component = shallow(
            <Cart {...props} cartItems={cartItems} totalPrice={productPrice} />
        );

        expect(component.find(".cart__title")).toHaveLength(1);
        expect(component.find(".cart__items")).toHaveLength(1);
        expect(component.find(".cart__footer")).toHaveLength(1);
    });
});
