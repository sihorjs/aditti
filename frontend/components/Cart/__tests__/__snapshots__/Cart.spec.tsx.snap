// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders without crash 1`] = `ShallowWrapper {}`;
